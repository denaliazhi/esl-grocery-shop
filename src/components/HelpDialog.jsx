/**
 * This component provides directions at the start
 * of the lesson. It's a dialog box that can be closed
 * or re-opened at any time.
 */
import { useState } from "react";

export default function HelpDialog({ closeText, children }) {
  const [showDialog, setShowDialog] = useState(true);

  function handleClick() {
    setShowDialog(!showDialog);
  }

  return (
    <>
      {showDialog ? (
        // Wrapper div to dim/blur content behind dialog box
        <div className="dialog-overlay">
          <dialog open>
            {children}
            <button className="close-help-btn" onClick={handleClick}>
              {closeText}
            </button>
          </dialog>
        </div>
      ) : (
        // Button to open dialog box
        <div>
          <button className="show-help-btn" onClick={handleClick}>
            Help
          </button>
        </div>
      )}
    </>
  );
}
