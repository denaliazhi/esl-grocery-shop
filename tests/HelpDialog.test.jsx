import { describe, test, expect, beforeEach } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import HelpDialog from "../src/components/HelpDialog";

describe("Help dialog", () => {
  test("when open, renders expected contents", () => {
    const instructions = "Follow these instructions";

    render(<HelpDialog closeText="Close">{instructions}</HelpDialog>);
    const dialog = screen.getByRole("dialog");

    expect(dialog).toBeInTheDocument();
    expect(dialog.textContent).toMatch(instructions);
  });

  test("when open, renders 'Close' not 'Help' button", () => {
    render(<HelpDialog closeText="Close">"Blah"</HelpDialog>);
    const closeButton = screen.getAllByRole("button");

    expect(closeButton).toHaveLength(1);
    expect(closeButton[0].textContent).toMatch("Close");
  });
});

describe("Help dialog | Interactions", () => {
  let user, closeButton;

  beforeEach(() => {
    user = userEvent.setup();
    render(<HelpDialog closeText="Close">"Blah"</HelpDialog>);
    closeButton = screen.getByRole("button");
  });

  test("clicking 'Close' button closes dialog", async () => {
    const dialog = screen.getByRole("dialog");
    await user.click(closeButton);

    expect(dialog).not.toBeInTheDocument();
  });

  test("when closed, renders 'Help' not 'Close' button", async () => {
    await user.click(closeButton);
    const helpButton = screen.getAllByRole("button");

    expect(helpButton).toHaveLength(1);
    expect(helpButton[0].textContent).toMatch("Help");
  });

  test("clicking 'Help' button opens dialog", async () => {
    await user.click(closeButton);
    const helpButton = screen.getByRole("button");
    await user.click(helpButton);
    const dialog = screen.getByRole("dialog");

    expect(dialog).toBeInTheDocument();
  });
});
