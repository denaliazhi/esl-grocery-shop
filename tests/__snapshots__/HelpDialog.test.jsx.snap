// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Help dialog > renders dialog box with expected contents 1`] = `
<div>
  <div
    class="dialog-overlay"
  >
    <dialog
      open=""
    >
      Follow these instructions
      <button
        class="close-help-btn"
      >
        Close
      </button>
    </dialog>
  </div>
</div>
`;

exports[`Help dialog > when open, renders expected contents 1`] = `
<div>
  <div
    class="dialog-overlay"
  >
    <dialog
      open=""
    >
      Follow these instructions
      <button
        class="close-help-btn"
      >
        Close
      </button>
    </dialog>
  </div>
</div>
`;
